File: Hash-Based Searching/Hopscotch Hashing Search/HopscotchHashingSearch.py
TABLE_SIZE = 16
HOP_RANGE = 4


class Bucket:

    def __init__(self):

        self.key = -1
        self.occupied = False
        self.hop_info = 0


class HopscotchHashingSearch:

    def __init__(self):

        self.table = [
            Bucket()
            for _ in range(TABLE_SIZE)
        ]

    def hash_function(self, key):

        return key % TABLE_SIZE

    def insert(self, key):

        home = self.hash_function(key)

        free_pos = -1

        for i in range(
            home,
            TABLE_SIZE
        ):

            if not self.table[i].occupied:

                free_pos = i
                break

        if free_pos == -1:
            return False

        if (
            free_pos - home
            >= HOP_RANGE
        ):
            return False

        self.table[
            free_pos
        ].key = key

        self.table[
            free_pos
        ].occupied = True

        self.table[
            home
        ].hop_info |= (
            1
            << (free_pos - home)
        )

        return True

    def search(self, key):

        home = self.hash_function(key)

        hop = self.table[
            home
        ].hop_info

        for i in range(HOP_RANGE):

            if hop & (1 << i):

                pos = home + i

                if (
                    pos < TABLE_SIZE
                    and
                    self.table[pos]
                    .occupied
                    and
                    self.table[pos]
                    .key == key
                ):

                    return True

        return False

    def display(self):

        print(
            "\nHopscotch Hash Table"
        )

        for i in range(TABLE_SIZE):

            if (
                self.table[i]
                .occupied
            ):

                value = (
                    self.table[i]
                    .key
                )

            else:

                value = "EMPTY"

            print(
                f"[{i}] = "
                f"{value} "
                f"(hop="
                f"{self.table[i].hop_info})"
            )


if __name__ == "__main__":

    ht =
        HopscotchHashingSearch()

    ht.insert(5)
    ht.insert(21)
    ht.insert(37)
    ht.insert(53)

    ht.display()

    print(
        "\nSearch 37:",
        "Found"
        if ht.search(37)
        else "Not Found"
    )

    print(
        "Search 100:",
        "Found"
        if ht.search(100)
        else "Not Found"
    )

File: Hash-Based Searching/Hopscotch Hashing Search/HopscotchHashingSearch.java
public class HopscotchHashingSearch {

    private static final int TABLE_SIZE = 16;
    private static final int HOP_RANGE = 4;

    static class Bucket {

        int key = -1;
        boolean occupied = false;
        int hopInfo = 0;
    }

    private final Bucket[] table;

    public HopscotchHashingSearch() {

        table =
            new Bucket[TABLE_SIZE];

        for (
            int i = 0;
            i < TABLE_SIZE;
            i++
        ) {

            table[i] =
                new Bucket();
        }
    }

    private int hashFunction(
        int key
    ) {

        return key % TABLE_SIZE;
    }

    public boolean insert(
        int key
    ) {

        int home =
            hashFunction(key);

        int freePos = -1;

        for (
            int i = home;
            i < TABLE_SIZE;
            i++
        ) {

            if (
                !table[i]
                .occupied
            ) {

                freePos = i;
                break;
            }
        }

        if (freePos == -1)
            return false;

        if (
            freePos - home >=
            HOP_RANGE
        )
            return false;

        table[
            freePos
        ].key = key;

        table[
            freePos
        ].occupied = true;

        table[
            home
        ].hopInfo |=
            (
                1 <<
                (
                    freePos
                    - home
                )
            );

        return true;
    }

    public boolean search(
        int key
    ) {

        int home =
            hashFunction(key);

        int hop =
            table[home]
            .hopInfo;

        for (
            int i = 0;
            i < HOP_RANGE;
            i++
        ) {

            if (
                (hop &
                (1 << i))
                != 0
            ) {

                int pos =
                    home + i;

                if (
                    pos <
                    TABLE_SIZE &&
                    table[pos]
                    .occupied &&
                    table[pos]
                    .key == key
                ) {

                    return true;
                }
            }
        }

        return false;
    }

    public void display() {

        System.out.println(
            "\nHopscotch "
            + "Hash Table"
        );

        for (
            int i = 0;
            i < TABLE_SIZE;
            i++
        ) {

            System.out.print(
                "[" + i +
                "] = "
            );

            if (
                table[i]
                .occupied
            ) {

                System.out.print(
                    table[i].key
                );
            }
            else {

                System.out.print(
                    "EMPTY"
                );
            }

            System.out.println(
                " (hop="
                + table[i]
                .hopInfo
                + ")"
            );
        }
    }

    public static void main(
        String[] args
    ) {

        HopscotchHashingSearch ht =
            new HopscotchHashingSearch();

        ht.insert(5);
        ht.insert(21);
        ht.insert(37);
        ht.insert(53);

        ht.display();

        System.out.println(
            "\nSearch 37: "
            + (
                ht.search(37)
                ? "Found"
                : "Not Found"
            )
        );
    }
}
